boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+		return u20, nil
+	}
+
+	// if kernel_status is trying, then we need to figure out what kernel we did
+	// boot from
+	sn, exists, err := u20.ebl.TryKernel()
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// somehow we booted with kernel_status=trying, but no try-kernel
+		// symlink :-/
+		return nil, fmt.Errorf("cannot mark boot successful: kernel_status is \"trying\", but no try-kernel enabled")
+	}
+
+	// set the snap we tried to the snap from the try-kernel symlink
+	u20.snapTried = sn
+
+	// update the boot var
+	u20.toCommit["kernel_status"] = ""
+
+	return u20, nil
+}
+
+func (s20 *bootState20Kernel) setNext(nextKernel snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	u20, err := newBootStateUpdate20(nil, snap.TypeKernel)
+	if err != nil {
+		return false, nil, err
+	}
+
+	currentKernel, err := u20.ebl.Kernel()
+	if err != nil {
+		return false, nil, err
+	}
+
+	kernelStatus := "try"
+	rebootRequired = true
+	// check if the next kernel snap is the same as the current kernel snap,
+	// in which case we either don't need to do anything or we just need to
+	// clear the status and not reboot
+	if nextKernel.SnapName() == currentKernel.SnapName() && nextKernel.SnapRevision() == currentKernel.SnapRevision() {
+		// If we were in anything but default ("") mode before
+		// and switched to the good core/kernel again, make
+		// sure to clean the kernel_status here. This also
+		// mitigates https://forum.snapcraft.io/t/5253
+		if u20.env["kernel_status"] == "" {
+			// already clean
+			return false, nil, nil
+		}
+
+		// clean
+		kernelStatus = ""
+		rebootRequired = false
+	} else {
+		// need to enable the symlink
+		u20.snapToTry = nextKernel
+	}
+
+	u20.toCommit["kernel_status"] = kernelStatus
+
+	return rebootRequired, u20, nil
+}
+
+type bootStateUpdate20 struct {
+	typ       snap.Type
+	ebl       bootloader.ExtractedRunKernelImageBootloader
+	env       map[string]string
+	toCommit  map[string]string
+	snapToTry snap.PlaceInfo
+	snapTried snap.PlaceInfo
+}
+
+func newBootStateUpdate20(u bootStateUpdate, typ snap.Type) (*bootStateUpdate20, error) {
+	// TODO:UC20: re-factor to handle base snap update as well as kernel snap
+	// TODO:UC20: handle base_status from modeenv as well
+	if u != nil {
+		u20, ok := u.(*bootStateUpdate20)
+		if !ok {
+			return nil, fmt.Errorf("internal error: threading unexpected boot state update with 20: %T", u)
+		}
+		return u20, nil
+	}
+	bl, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, err
+	}
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if !ok {
+		return nil, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
+	}
+
+	m, err := bl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, err
+	}
+	return &bootStateUpdate20{
+		ebl:      ebl,
+		env:      m,
+		toCommit: make(map[string]string),
+		typ:      typ,
+	}, nil
+}
+
+func (u20 *bootStateUpdate20) commit() error {
```

Comment:
Feels like it's a bit of special casing in the code here, there's effectively 2 operations here, marking the kernel as successfully booted, o trying a new update. Perhaps it'd be nicer to have separate methods, each for one scenario. Maybe drop `toCommit` at that point too.

Reply:
Forgot to mention, this is a follow up material too.

Reply:
Yes I think a refactoring of this whole file is due when I add base snap updates, it's getting a bit duplicative and I think with the right organization this can be cleaned up to do what you say as well.